Documentation for temp table and boolean support (#537)
tableau/connector-plugin-sdk: docs/dialect.md

  **Join Capabilities**  <br/>
  Join usage is also defined by capabilities in the manifest file. See the Query section for join relation capabilities [here]({{ site.baseurl }}/docs/capabilities).

### Boolean Support:
  Some databases need to customize boolean support functions.  A common case is when a database lacks native boolean support. <br/>
  `format-true` and `format-false` are used in predicate statements. <br/>
   **format-true** <br/> 
  ```xml 
   <format-true value='(1=1)' />
  ```
  **format-false** <br/>
  ```xml 
  <format-false value='(1=0)' />
  ```

**Note: Changes in Tableau 2020.3**

  With the release of Tableau 2020.3, `format-true` and `format-false` will have two parameters, literal and predicate. <br/>
  Literals is used when a boolean is used as a value. Default to: `x ? 1 : 0` <br/>
  Example usage: `SELECT 1 as literalBool` <br/>

>   Predicates is used when a boolean is used as a predicate. Default to: `isTrue ? (1=1) : (1=0)` <br/>

Reviewer: The wording here is a little confusing. Since there are no `x ? 1 : 0` or `isTrue ? (1=1) : (1=0)` strings anywhere in our code as far as I know, try saying "the default values are 1 and 0" and "the default values are '(1=1)' and '(1=0)'", respectively.

Author: Got it. This seems more clear.

>   Example usage: `SELECT something as test WHERE (1=1)` <br/>

Reviewer: If you're capitalizing keywords, capitalize AS, too, for consistency

Author: Made this change.

  `value` attribute is used as `predicate` for backwards compatability <br/>
  Example:
  ```xml 
    <format-true literal='TRUE' predicate='TRUE' />
  ```
  ```xml 
    <format-false literal='FALSE' predicate='FALSE' />
  ```

   **format-bool-as-value** <br/>

>     Used in CASE statements. Determines whether the true or false case is used first. The fucntion is only used when `CAP_QUERY_BOOLEXPR_TO_INTEXPR` capiblity is set to yes. 

Reviewer: "The function is only used when _the_ `CAP_QUERY_BOOLEXPR_TO_INTEXPR` _capability_ is set to yes."

Author: Fixed this.

  ```xml 
    <format-bool-as-value value='TrueFirst' />
  ```
  
 **Boolean Capabilities**  <br/>
  
  Boolean usage is also defined by capabilities in the manifest file. <br\>
  See the Query section for boolean capabilities [here]({{ site.baseurl }}/docs/capabilities) for more details.

 **TDVT Coverage**  <br/>
The following TDVT tests check that the boolean functionality is working as expected for a connector. 
logical.bool	`exprtests\standard\setup.logical.bool.txt` and logical	`exprtests\standard\setup.logical.txt`. See [this](https://github.com/tableau/connector-plugin-sdk/tree/master/tdvt/tdvt/exprtests/standard) for more details. 

### Temporary Table Support:
   **format-create-table** <br/> 
  This function uses  piece-by-piece formulas for creating a table.
  Predicates can be used with tokens that only corresponds to a certain type of table.
  Predicates: <br/>
  -  __GlobalTemp__
  -  __LocalTemp__
  -  __AnyTemp__
  -  __NoTemp__ <br/>
   You should also be able to use the following string substitution tokens along with the predicate:
      - __%n - table name__
     - __%f - formatted column list__

>     This function is avialable when the `CAP_CREATE_TEMP_TABLES` capability is set to yes. 

Reviewer 2: ```suggestion
    This function is available when the `CAP_CREATE_TEMP_TABLES` capability is set to yes. 
```

  
  ```xml 
    <format-create-table>
        <formula>CREATE </formula>
        <formula predicate='GlobalTemp'>GLOBAL </formula>
        <formula predicate='LocalTemp'>LOCAL </formula>
        <formula predicate='AnyTemp'>TEMPORARY </formula>
        <formula>TABLE %n (</formula>
        <formula>%f</formula>
        <formula>)</formula>
        <formula predicate='AnyTemp'> ON COMMIT PRESERVE ROWS</formula>
      </format-create-table>
  ```

  **format-select** <br/>

>   This function uses piece-by-piece formula for defining a SELECT statement. Here, we can define the clause used in `SELECT` statement. <br/>

Reviewer: Grammar nit: "This function uses _a_ piece-by-piece formula..."
"Here, we can define the _clauses_ used in _the_ `SELECT` statement."

  the `Into` clause in `SELECT` statement creates a new table. `<format-select>` will help you define how your TEMP table is created when using `INTO` clause.

>   `INTO` cluse is only aviable when `CAP_SELECT_INTO` capability is set to yes. 

Reviewer: More minor grammer/spelling nits:
"the `Into` clause in _the_ `SELECT` statement..."
"...is created when using _an_ `INTO` clause."
"`INTO` _clause_ is only available when _the_ `CAP_SELECT_INTO` capability..."

>   `INTO` cluse is only aviable when the `CAP_SELECT_INTO` capability is set to yes. 

Reviewer 2: ```suggestion
  `INTO` clause is only available when the `CAP_SELECT_INTO` capability is set to yes. 
```

  
  ```xml 
      <format-select>
        <part name='Into' value='CREATE GLOBAL TEMPORARY TABLE %1 ON COMMIT PRESERVE ROWS AS' />
        <part name='Top' value='SELECT * FROM (' />
        <part name='Select' value='SELECT %1' />
        <part name='From' value='FROM %1' />
        <part name='Where' value='WHERE %1' />
        ...
      </format-select> />
  ```

  **format-drop-table** <br/>
    This function defines the format for dropping a table. %1 is the table name. Each formula is executed as a separate statement. <br/>
    This function is only used when `CAP_TEMP_TABLES_NOT_SESSION_SCOPED` capability is set to yes. 
  ```xml 
    <format-drop-table>

>       <formula>TRUNCATE TABLE %1</formula>
>       <formula>DROP TABLE %1 PURGE</formula>

Reviewer 2: ```suggestion
      <formula>DROP TABLE %1 </formula>
```
The other code works but I think this example is more clear for a doc.

    </format-drop-table>
  ``` 
  

>  **Temp Tables Capabilities**  <br/>

Reviewer: Nit: Should probably be singular like the other capabilities sections, i.e. "Temp Table Capabilities"

Author: Made this change. Makes sense.

  **Join Capabilities**  <br/>
  Join usage is also defined by capabilities in the manifest file. See the Query section for join relation capabilities [here]({{ site.baseurl }}/docs/capabilities).

### Boolean Support:

>   We can use boolean support functions when a database lacks native boolean support. <br/>

Reviewer 2: ```suggestion
  Some databases need to customize boolean support functions.  A common case is when a database lacks native boolean support. <br/>
```

Author: Seems clear made the changes.

>    **format-true** <br/> 

Reviewer 3: @rosswbrown, didn't you change how this works in 2020.3? We should probably have both paths here if so.

Reviewer 2: Yes, there should be a "Tableau 2020.3 and later" which uses the `literal` and `predicate` form.  There should also be the general documentation for 'value' which was the only extension point pre-2020.3.  Details in the CL2186161 description.

  String literal or expression for boolean false value.  
  ```xml 
   <format-true value='(1=1)' />
  ```
  **format-false** <br/>
  String literal or expression for boolean false value. 
  ```xml 
  <format-false value='(1=0)' />
  ```

>    **format-bool-as-value** <br/>

Reviewer 2: This is only used when CAP_QUERY_BOOLEXPR_TO_INTEXPR is yes

Author: Removed this part and kept only the link to the details for capabilities

    Used in CASE statements. Determines whether the true or false case is used first. 
    The value for this property can be: 
  - __TrueFirst__
      This is the default case. Logic: `CASE WHEN %1 THEN 1 WHEN NOT %1 THEN 0 ELSE NULL END`
  - __FalseFirst__
      False case is used first. Logic: `CASE WHEN NOT %1 THEN 0 WHEN %1 THEN 1 ELSE NULL END`

  ```xml 
    <format-bool-as-value value='TrueFirst' />
  ```
  
 **Boolean Capabilities**  <br/>
  
  Boolean usage is also defined by capabilities in the manifest file. The capabilities are: <br>
  - __CAP_QUERY_BOOLEXPR_TO_INTEXPR__

>   -  __CAP_QUERY_BOOLEXPR_TO_INTEXPR__

Reviewer 2: ```suggestion
```
duplicate

Author: Removed this part and kept only the link to the details for capabilities

  - __CAP_QUERY_BOOL_IDENTIFIER_TO_LOGICAL__
  - __CAP_QUERY_GROUP_BY_BOOL__
  - __CAP_ODBC_BIND_BOOL_AS_WCHAR_TFLITERAL__ <br/>
  This capability seems to be necessary only for Hive/Impala connector. 

> - __CAP_ODBC_BIND_BOOL_AS_WCHAR_01LITERAL__ <br/> 

Reviewer 2: I think these actually confuse the issue.  Let's stick to the query side stuff for this doc.

Author: Removed this part and kept only the link to the details for capabilities

  See the Query section for boolean capabilities [here]({{ site.baseurl }}/docs/capabilities) for more details.

 **TDVT Coverage**  <br/>

> The following TDVT tests check that the boolean functions are working as expected in tabquery for a database connector. 

Reviewer 2: ```suggestion
The following TDVT tests check that the boolean functionality is working as expected for a connector. 
```

Author: Made these changes, the language seems more clear.

logical.bool	`exprtests\standard\setup.logical.bool.txt` and logical	`exprtests\standard\setup.logical.txt`. See [this](https://github.com/tableau/connector-plugin-sdk/tree/master/tdvt/tdvt/exprtests/standard) for more details. 

> ### Temp Table Support:

Reviewer 2: ```suggestion
### Temporary Table Support:
```

Author: This seems more suitable for the documentation.

   **format-create-table** <br/> 

>   This function uses  piece-by-piece formulas for creating a table.

Reviewer 2: This is controlled by CAP_CREATE_TEMP_TABLES = yes

Author: Added this detail to the documentation.

  Predicates can be used with tokens that only corresponds to a certain type of table.
  Predicates: <br/>
  -  __GlobalTemp__
  -  __LocalTemp__
  -  __AnyTemp__
  -  __NoTemp__ <br/>
   You should also be able to use the following string substitution tokens along with the predicate:
      - __%n - table name__
     - __%f - formatted column list__
      
  ```xml 
  <format-create-table>
      <formula>CREATE </formula>
      <formula predicate='GlobalTemp'>GLOBAL </formula>
      <formula predicate='LocalTemp'>LOCAL </formula>
      <formula predicate='AnyTemp'>TEMPORARY </formula>
      <formula>TABLE %n (</formula>
      <formula>%f</formula>
      <formula>)</formula>
      <formula predicate='AnyTemp'> ON COMMIT PRESERVE ROWS</formula>
    </format-create-table>
  ```
  **format-select** <br/>

>     This function uses piece-by-piece formula for defining a SELECT statement. Here, we can define the clause used in `SELECT` statement. <br/>

Reviewer 2: I think this, with Into, will only be used CAP_SELECT_INTO = yes

Author: Added this detail to the documentation.

    the `Into` clause in `SELECT` statement creates a new table. `<format-select>` will help you define how your TEMP table is created when using `INTO` clause.
  ```xml 
    <format-select>
      <part name='Into' value='CREATE GLOBAL TEMPORARY TABLE %1 ON COMMIT PRESERVE ROWS AS' />
      <part name='Top' value='SELECT * FROM (' />
      <part name='Select' value='SELECT %1' />
      <part name='From' value='FROM %1' />
      <part name='Where' value='WHERE %1' />
      ...
    </format-select> />
  ```

>   

Reviewer 2: We need a section about `<format-drop-table>` if CAP_TEMP_TABLES_NOT_SESSION_SCOPED = yes

Author: Updated the changes to reflect format drop table.

  **Join Capabilities**  <br/>
  Join usage is also defined by capabilities in the manifest file. See the Query section for join relation capabilities [here]({{ site.baseurl }}/docs/capabilities).

### Boolean Support:
  Some databases need to customize boolean support functions.  A common case is when a database lacks native boolean support. <br/>
   **format-true** <br/> 

>   String literal or expression for boolean false value.  

Reviewer 2: Only predicate customization supported pre-2020.3.  Remove mention of literal in both format-true and format-false

Author: Removed the mention of literal support for  pre-2020.3

Author: ![boolean_support](https://user-images.githubusercontent.com/14967333/85333338-44273f80-b48e-11ea-9303-413e7ad6e945.PNG)
![temp_tables](https://user-images.githubusercontent.com/14967333/85333340-44bfd600-b48e-11ea-87d6-de46ba847104.PNG)

  ```xml 
   <format-true value='(1=1)' />
  ```

> 'format-true' and 'format-false' are used in two use cases: literals and predicates

Reviewer 2: Move this above format-true and update it to use 
```suggestion
'format-true' and 'format-false' are used in predicate statements
```

Author: Removed the mention of literal support and added `'format-true' and 'format-false' are used in predicate statements` in the documentation
